File: src/langarian/receipts.py
"""Operation receipts and stable hashing."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any
import hashlib
import json
import os
import tempfile

from .claims import Claim
from .contracts import InvariantResult
from .epistemic import EpistemicTag, ResultStatus, combine_statuses
from .version import KERNEL_VERSION, METRIC_VERSION, RECEIPT_SCHEMA_VERSION


def canonical_json(data: Any) -> str:
    return json.dumps(data, sort_keys=True, separators=(",", ":"), ensure_ascii=False, default=str)


def _sha256_json(data: Any) -> str:
    return "sha256:" + hashlib.sha256(canonical_json(data).encode("utf-8")).hexdigest()


def atomic_write_text(path: str | Path, text: str) -> Path:
    """Write ``text`` to ``path`` atomically (tmp file in the same directory + rename).

    A crash mid-write can never leave a truncated receipt at ``path``; the
    temporary file is cleaned up on failure. SPEC section 3.11.
    """

    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp_name = tempfile.mkstemp(prefix=path.name + ".", suffix=".tmp", dir=path.parent)
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            handle.write(text)
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(tmp_name, path)
    except BaseException:
        try:
            os.unlink(tmp_name)
        except OSError:
            pass
        raise
    return path


@dataclass(frozen=True)
class OperationReceipt:
    """Immutable record of one kernel operation.

    Two distinct identities are provided:

    - ``content_hash()``: deterministic mathematical identity of the
      operation record. It excludes ``timestamp_utc`` and ``receipt_id``, so
      identical operations at any time produce the same content hash. Lineage
      uses this for mathematical identity.
    - ``receipt_id()``: emission-event identity. It includes ``timestamp_utc``,
      so each emission is unique even for identical operations. Ledgers use
      this for audit events. It is not a content hash.
    """

    operator: str
    input_hashes: list[str]
    output_hash: str
    parameters: dict[str, Any] = field(default_factory=dict)
    coherence_before: float | None = None
    coherence_after: float | None = None
    invariant_results: list[InvariantResult] = field(default_factory=list)
    epistemic_tag: EpistemicTag = EpistemicTag.COMPUTED
    claims: list[Claim] = field(default_factory=list)
    timestamp_utc: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())

    @property
    def status(self) -> ResultStatus:
        if self.epistemic_tag == EpistemicTag.FAILED:
            return ResultStatus.FAIL
        return combine_statuses([result.status for result in self.invariant_results])

    def body(self, include_receipt_id: bool = False) -> dict[str, Any]:
        body = {
            "kernel_version": KERNEL_VERSION,
            "metric_version": METRIC_VERSION,
            "receipt_schema_version": RECEIPT_SCHEMA_VERSION,
            "timestamp_utc": self.timestamp_utc,
            "operator": self.operator,
            "input_hashes": self.input_hashes,
            "output_hash": self.output_hash,
            "parameters": self.parameters,
            "coherence_before": self.coherence_before,
            "coherence_after": self.coherence_after,
            "invariant_results": [result.to_dict() for result in self.invariant_results],
            "status": self.status.value,
            "epistemic_tag": self.epistemic_tag.value,
            "claims": [claim.to_dict() for claim in self.claims],
        }
        if include_receipt_id:
            body["content_hash"] = self.content_hash()
            body["receipt_id"] = self.receipt_id()
        return body

    def content_hash(self) -> str:
        """Deterministic content identity; excludes timestamp and receipt_id."""

        body = self.body(include_receipt_id=False)
        body.pop("timestamp_utc", None)
        return _sha256_json(body)

    def receipt_id(self) -> str:
        """Emission-event identity; includes timestamp_utc (unique per emission)."""

        return _sha256_json(self.body(include_receipt_id=False))

    def to_json(self) -> str:
        return json.dumps(self.body(include_receipt_id=True), indent=2, sort_keys=True, ensure_ascii=False, default=str)

    def write_json(self, path: str) -> None:
        atomic_write_text(path, self.to_json() + "\n")

File: web/src/kernel/receipts.ts
/**
 * Operation receipts and stable hashing (TS mirror of src/langarian/receipts.py).
 *
 * Two distinct identities:
 * - contentHash(): deterministic mathematical identity of the operation
 *   record; excludes timestamp_utc and receipt_id.
 * - receiptId(): emission-event identity; includes timestamp_utc.
 */

import { canonicalJson, PyFloat, type CanonicalValue } from './canonical.js'
import { combineStatuses, type InvariantResult, type ResultStatus } from './contracts.js'
import { sha256Prefixed } from './sha256.js'
import { KERNEL_VERSION, METRIC_VERSION, RECEIPT_SCHEMA_VERSION } from './version.js'

export type EpistemicTag = 'FORMAL' | 'COMPUTED' | 'MODEL' | 'INTERPRETIVE' | 'METAPHOR' | 'OBSERVED' | 'FAILED'

export interface Claim {
  text: string
  tag: EpistemicTag
  evidence: readonly string[]
  metadata: Record<string, CanonicalValue>
}

export const claim = (text: string, tag: EpistemicTag): Claim => ({ text, tag, evidence: [], metadata: {} })

export interface OperationReceiptInit {
  operator: string
  inputHashes: readonly string[]
  outputHash: string
  parameters?: Record<string, CanonicalValue>
  coherenceBefore?: number | null
  coherenceAfter?: number | null
  invariantResults?: readonly InvariantResult[]
  epistemicTag?: EpistemicTag
  claims?: readonly Claim[]
  timestampUtc?: string
}

function invariantToDict(result: InvariantResult): Record<string, CanonicalValue> {
  return {
    name: result.name,
    status: result.status,
    message: result.message,
    value: result.value,
    metadata: result.metadata,
  }
}

function claimToDict(item: Claim): Record<string, CanonicalValue> {
  return {
    text: item.text,
    tag: item.tag,
    evidence: Array.from(item.evidence),
    metadata: item.metadata,
  }
}

export class OperationReceipt {
  readonly operator: string
  readonly inputHashes: readonly string[]
  readonly outputHash: string
  readonly parameters: Record<string, CanonicalValue>
  readonly coherenceBefore: number | null
  readonly coherenceAfter: number | null
  readonly invariantResults: readonly InvariantResult[]
  readonly epistemicTag: EpistemicTag
  readonly claims: readonly Claim[]
  readonly timestampUtc: string

  constructor(init: OperationReceiptInit) {
    this.operator = init.operator
    this.inputHashes = Object.freeze(init.inputHashes.slice())
    this.outputHash = init.outputHash
    this.parameters = Object.freeze({ ...(init.parameters ?? {}) })
    this.coherenceBefore = init.coherenceBefore ?? null
    this.coherenceAfter = init.coherenceAfter ?? null
    this.invariantResults = Object.freeze((init.invariantResults ?? []).slice())
    this.epistemicTag = init.epistemicTag ?? 'COMPUTED'
    this.claims = Object.freeze((init.claims ?? []).slice())
    this.timestampUtc = init.timestampUtc ?? new Date().toISOString()
    Object.freeze(this)
  }

  get status(): ResultStatus {
    if (this.epistemicTag === 'FAILED') return 'FAIL'
    return combineStatuses(this.invariantResults.map((result) => result.status))
  }

  body(includeReceiptId = false): Record<string, CanonicalValue> {
    const body: Record<string, CanonicalValue> = {
      kernel_version: KERNEL_VERSION,
      metric_version: METRIC_VERSION,
      receipt_schema_version: RECEIPT_SCHEMA_VERSION,
      timestamp_utc: this.timestampUtc,
      operator: this.operator,
      input_hashes: Array.from(this.inputHashes),
      output_hash: this.outputHash,
      parameters: this.parameters as CanonicalValue,
      coherence_before: this.coherenceBefore === null ? null : new PyFloat(this.coherenceBefore),
      coherence_after: this.coherenceAfter === null ? null : new PyFloat(this.coherenceAfter),
      invariant_results: this.invariantResults.map(invariantToDict),
      status: this.status,
      epistemic_tag: this.epistemicTag,
      claims: this.claims.map(claimToDict),
    }
    if (includeReceiptId) {
      body.content_hash = this.contentHash()
      body.receipt_id = this.receiptId()
    }
    return body
  }

  /** Deterministic content identity; excludes timestamp and receipt_id. */
  contentHash(): string {
    const body = this.body(false)
    delete body.timestamp_utc
    return sha256Prefixed(canonicalJson(body))
  }

  /** Emission-event identity; includes timestamp_utc (unique per emission). */
  receiptId(): string {
    return sha256Prefixed(canonicalJson(this.body(false)))
  }

  /** Full receipt body including content_hash and receipt_id. */
  toBody(): Record<string, CanonicalValue> {
    return this.body(true)
  }

  /** Canonical JSON of the full body (byte-exact with the Python kernel). */
  toCanonicalJson(): string {
    return canonicalJson(this.toBody())
  }
}
